File: Sean的leetcode/Leetcode/Create Maximum Number.cpp
/* DP, greedy
 * maxNumberLimit: find the max number of a limited length, keep the order decreasing, pop the last element if smaller
 * merge: merge two vectors to form a max number
 * greater: compare two vector of number
 */

class Solution {
public:
    vector<int> maxNumber(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<int> greatest;
        int n = nums1.size();
        int m = nums2.size();
        for (int l1 = max(0, k - m); l1 <= min(n, k); l1++) {
            greatest = greater(greatest, merge(maxNumberLimit(nums1, l1), maxNumberLimit(nums2, k - l1)));
        }
        return greatest;
    }
    vector<int> maxNumberLimit(vector<int>& nums, int l) {
        int drop = nums.size() - l;
        vector<int> res;
        for (auto num : nums) {
            while (drop && !res.empty() && res.back() < num) {
                res.pop_back();
                drop--;
            }
            res.push_back(num);
        }
        res.resize(l)
        return res;
    }
    vector<int> merge(vector<int> nums1, vector<int> nums2) {
        int n = nums1.size();
        int m = nums2.size();
        int i = 0, j = 0;
        vector<int> res;
        while (i < n || j < m) {
            if (i == n) {
                res.push_back(nums2[j++]);
            }
            else if (j == m) {
                res.push_back(nums1[i++]);
            }
            else {
                if (nums1[i] == nums2[j]) {
                    int k = i, l = j;
                    while (k < n && l < m && nums1[k] == nums2[l]) {
                        k++;
                        l++;
                    }
                    if (l == m || k < n && nums1[k] > nums2[l]) {
                        res.push_back(nums1[i++]);
                    }
                    else {
                        res.push_back(nums2[j++]);
                    }
                }
                else if (nums1[i] > nums2[j]) {
                    res.push_back(nums1[i++]);
                }
                else {
                    res.push_back(nums2[j++]);
                }
            }
        }
        return res;
        
    }
    vector<int> greater(vector<int> nums1, vector<int> nums2) {
        int n = nums1.size();
        int m = nums2.size();
        if (n != m) return n > m? nums1 : nums2;
        int i = 0;
        while (i < n) {
            if (nums1[i] != nums2[i]) {
                return nums1[i] > nums2[i]? nums1 : nums2;
            }
            i++;
        }
        return nums1;
    }
};

File: Sean的leetcode/python/Create Maximum Number.py
###
# Array Stack
# Time Complexity: k * [O(n+m)+O(k)]      
###
class Solution(object):
    def maxNumber(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[int]
        """
        re = []
        for l1 in xrange(max(0, k-len(nums2)), min(len(nums1), k)+1):
            re = max(re, self.merge(self.getmax(nums1, l1), self.getmax(nums2, k-l1)))
        return re
    
    def getmax(self, nums, l):
        rm = len(nums) - l
        re = []
        for n in nums:
            while re and rm and re[-1] < n:
                re.pop()
                rm -= 1
            re.append(n)
        return re[:l]
    
    def merge(self, nums1, nums2):
        re = []
        i = 0
        j = 0
        while i < len(nums1) or j < len(nums2):
            if i == len(nums1):
                re.append(nums2[j])
                j += 1
            elif j == len(nums2):
                re.append(nums1[i])
                i += 1
            else:
                if nums1[i] > nums2[j]:
                    re.append(nums1[i])
                    i += 1
                elif nums2[j] > nums1[i]:
                    re.append(nums2[j])
                    j += 1
                else:
                    k, l = i, j
                    while k < len(nums1) and l < len(nums2) and nums1[k] == nums2[l]:
                        k += 1
                        l += 1
                    if l == len(nums2) or (k < len(nums1) and nums1[k] > nums2[l]):
                        re.append(nums1[i])
                        i += 1
                    else:
                        re.append(nums2[j])
                        j += 1             
        return re
    
    
        